### ISSUE TITLE
CKEditor 5 – Roadmap to 1.0.0 and beyond

### ISSUE THREAD

**Issue by Reinmar** [ckeditor/ckeditor5-design#172]
The roadmap is divided into 3 parts:

* Done – topics which we already finished and done and consider ready to ship (may need bug-fixing and minor polishing),
* In progress – topics which we’re currently working on or which we started, and don’t necessarily work at the moment, if they are in an advanced stage,
* To do:
  * for 1.0.0 alpha,
  * for 1.0.0 final,
  * after 1.0.0.

We’ll be updating the roadmap to match the progress that we make.

## Done

* Editing engine – https://github.com/ckeditor/ckeditor5-engine.
  * Data model with support for Operational Transformation.
  * Rendering – view layer, model to view and view to model conversion, view rendering (structure + selection).
  * Editing and data controllers.
  * Set of generic utils to work with the model.
  * Set of test tools (to speed up writing automated tests).
* Editor core infrastructure – https://github.com/ckeditor/ckeditor5-engine.
  * Base editor classes.
  * Commands API.
  * Plugin system.
* UI library with a set of basic components and a wireframe theme (base for other themes) – https://github.com/ckeditor/ckeditor5-ui.
  * Simple MVC library.
  * Basic components like button, dropdown, toolbar, floating balloon, input.
  * Set of composable utils like form and focus cycle managers.
* Utils library – https://github.com/ckeditor/ckeditor5-utils.
* Undo feature – https://github.com/ckeditor/ckeditor5-undo.
* Paragraph feature – https://github.com/ckeditor/ckeditor5-paragraph.
* Enter feature – https://github.com/ckeditor/ckeditor5-enter.
* Typing feature – https://github.com/ckeditor/ckeditor5-typing.
* Clipboard integration (copy, cut, paste) – https://github.com/ckeditor/ckeditor5-clipboard.
* Heading feature – https://github.com/ckeditor/ckeditor5-heading.
* Image feature – https://github.com/ckeditor/ckeditor5-image.
  * Block image support using widgets.
  * Alternate text panel.
  * Image styles.
* Link feature – https://github.com/ckeditor/ckeditor5-link.
* List feature – https://github.com/ckeditor/ckeditor5-list
  * Only flat lists are supported so far (nested weren't completed).
* Basic styles feature (bold and italic) – https://github.com/ckeditor/ckeditor5-basic-styles.
* First theme – https://github.com/ckeditor/ckeditor5-theme-lark.
* Autoformatting feature – https://github.com/ckeditor/ckeditor5-autoformat.
* Classic editor creator – https://github.com/ckeditor/ckeditor5-creator-classic.
* Support for Chrome, Opera, Safari, Firefox (52+).

## In progress

* List feature – https://github.com/ckeditor/ckeditor5-list.
  * Nested lists still require some work.
* Image feature completeness – https://github.com/ckeditor/ckeditor5-image.
  * Captioning support.
* Inline editor creator – https://github.com/ckeditor/ckeditor5-creator-inline.
* Translation service – https://github.com/ckeditor/ckeditor5/issues/387.
* Project infrastructure review – best described in https://github.com/ckeditor/ckeditor5-design/issues/171.
* Widget system (actually – a small set of utils, because the necessary basics are part of the engine).
* Basic accessibility support (keyboard navigation + basic ARIA attributes).

### For 1.0.0 alpha

* Blockquote feature.
* Compatibility with MS Edge.
* Basic guides, tutorials and samples.
* Define official editor presets (a preset is a set of features).
* Schema refactor – https://github.com/ckeditor/ckeditor5-engine/issues/532.
* Tool for checking the whole project out to a past state (e.g. for bisecting). Needed because of the multi-repo architecture.
* Pasting stability improvements (i.e. stable content conversion forgiveness).
* Undo size limit and model garbage collection.
* Image dropping and pasting support.
* Cleanup on editor destruction (destruction chains) – https://github.com/ckeditor/ckeditor5/issues/114.
* Improved input methods (IME, accents, native MacOS tooltips).
* Critical bugfixes.
* Critical code cleanup and refactoring.

### For 1.0.0 final

* Review of the API documentation.
* Bugfixes.
* Code cleanup and refactoring (changes necessary for API stabilization).

### After 1.0.0

* Mobile (mainly iOS due to its very unpleasant viewport system):
  * Dedicated UI or improvements for the existing  – https://github.com/ckeditor/ckeditor5-design/issues/149.
  * UX polishing (for both, the editing and UI).
* Text transformation feature (“--” to mdash, quotes, disallow subsequent spaces, …, etc.).
* Autolinking.
* Complete drag and drop support:
  * for editor content (including widgets),
  * for content from outside.
* Paste from:
  * MS Word.
  * Google Docs.
* Autocompleter feature (e.g. mentions).
* Full accessibility audit and support (for screen readers).
* Image feature improvements:
  * Linking.
  * Image insert panel – https://github.com/ckeditor/ckeditor5-image/issues/16.
* More editor creators (types – e.g. headless editor, chromeless editor, etc.).
* IE11 support.
* Performance improvements.
* Magicline (as an API to be used by e.g. drag and drop feature and a standalone plugin).
* Ability to type after a link. Depending on a solution it can be a general ability to control selection gravity (like in Firefox) or some hack specific for links.
* Years of polishing the software ;)

**Comment by Reinmar** [comment]
Predicted ETA for 1.0.0 alpha is *early Q2 2017*.

The scope of 1.0.0 alpha is pretty well defined now and we just don't know how long some of these things can still take. Even with the alpha version we'd like to have a quite stable software and documentation which is complete enough to start tooling with the editor.

Between the alpha and final versions we'll be working on stability, code cleanup refactoring (the goal is to get as stable API after 1.0.0 final as possible) and additional resources.

Then... we're back to features :).

**Comment by Reinmar** [comment]
I updated the readme to reflect recent progress.

* Image captioning, translation service, inline editor creator, basic accessibility, final project infrastructure review and editor presets were moved to "Done".
* Blockquote, the tool for bisecting the multi-repo project and input methods stability were moved to "In progress".
* Also, I added a point about editor builds and placeholders support in the "In progress" section.

**Comment by Reinmar** [comment]
I update the first post to reflect the recent progress and decisions.

* Widget system, placeholders, editor builds, revision log, nested lists were moved to "Done".
* Garbage collector and schema refactor were moved to 1.0.0 final.
* The "In progress" section were updated.

**Comment by Reinmar** [comment]
I update the first post to reflect the recent progress and decisions.

* Image upload, content filtering stability, MS Edge support were moved to "Done".
* All what was left in "TODO / For 1.0.0 alpha" is now in progress :).

**Comment by maikaruaiphone** [comment]
Good morning, 
We you know when Final or possibly even BETA will be available? My company has a policy to wait until those phases, and thus we cannot use ALPHA even though I am ecstatic to hear these new updates in 5.0!

**Comment by Reinmar** [comment]
Hi! Good morning (or evening ;))!

The first beta release will be in Q1 2018. Perhaps even within a month, but I don't want to make promises. We're right now finishing the huge refactoring that we had planned before the first beta and are switching to polishing the release.

**Comment by adriancamara** [comment]
@Reinmar  - any update on the timing of this first beta release?

**Comment by Reinmar** [comment]
We finished the refactoring and we're happy with the results. We focus now on some minor cleanup, bug fixing and docs. Which means that there's little that can go wrong, so we'll release the beta in March. Right now targetting the middle of March.

**Comment by adriancamara** [comment]
@Reinmar will that include support for paste from MS Word?

**Comment by Reinmar** [comment]
No, PFW (paste from MS Word) and tables will be added later. We started working on tables recently and are going to start work on PFW soon.

**Comment by maikaruaiphone** [comment]
Happy Friday, Does it still seem like Middle of March as BETA release date? 
Thank you

**Comment by Reinmar** [comment]
Yes :)

**Comment by maikaruaiphone** [comment]
Thank you! This is good to hear

**Comment by maikaruaiphone** [comment]
Good morning, Wanted to check in and see how things were with the Mid March BETA Release?

Thank you :-)

**Comment by Reinmar** [comment]
The status of things:

![image](https://user-images.githubusercontent.com/156149/37291977-29aa502a-2610-11e8-94bf-3c91ebe340e0.png)

Besides docs (for which we have a never-ending list of tickets), we've got a couple of things left. All on review or in advanced progress. Wed/Thu this week is still our target :)

**Comment by maikaruaiphone** [comment]
Good morning Reinmar!
Hope your week is going well.

Wanted to check in and see if we should expect today or tomororw for BETA?

Thank you so much!

**Comment by Reinmar** [comment]
![image](https://user-images.githubusercontent.com/156149/37409935-d1752290-279f-11e8-8f9f-bc9ef0116888.png)

Tomorrow :) We've just finished a testing phase and we haven't found anything blocking.

**Comment by Reinmar** [comment]
The release is out 🎉 

![image](https://user-images.githubusercontent.com/156149/37465753-584f0e68-285c-11e8-9156-b4804711a33e.png)

It's tagged and it's published npm. We'll be publishing release notes, docs, CDN and what else later today and tomorrow. We're releasing CKEditor 5, CKEditor 4.9 and more today, so it's a busy time :)

**Comment by maikaruaiphone** [comment]
Good day, 

Thank you for this update! Where would we go to download and deploy this?

Thank you!

**Comment by Reinmar** [comment]
Just follow https://docs.ckeditor.com/ckeditor5/latest/builds/guides/integration/installation.html. And in case of issues, please report them on https://github.com/ckeditor/ckeditor5 to keep this thread clean.

**Comment by maikaruaiphone** [comment]
Good morning and thank you for your reply!
When i click the first link (https://docs.ckeditor.com/ckeditor5/latest/builds/guides/integration/installation.html then I click https://ckeditor.com/ckeditor-5-builds/download/) you provided I am still only seeing Alpha and not BETA anywhere.

**Comment by Reinmar** [comment]
Nope, you're right. We were meant to update the website today, but we also released [CKEditor 4.9](https://ckeditor.com/blog/CKEditor-4.9-released/) and [Easy Image for CKEditor Cloud Services](https://ckeditor.com/ckeditor-cloud-services/easy-image/) yesterday and there's a queue of pending updates. We should have the download sections and demos updated on https://ckeditor.com on Monday.

In the meantime, if you want to download beta.1 zip packages, you can do that from http://download.cksource.com/CKEditor/CKEditor5/1.0.0-beta.1/

**Comment by Reinmar** [comment]
We published release notes for beta.1+beta.2 in https://github.com/ckeditor/ckeditor5-design/issues/184.

And I updated the roadmap to best reflect the current status of things.

**Comment by Reinmar** [comment]
Finally! The first stable version of CKEditor 5 is out!

* Blog post: https://ckeditor.com/blog/CKEditor-5-v10.0.0-the-future-of-rich-text-editing-looks-stable/
* Release notes: https://github.com/ckeditor/ckeditor5/releases/tag/v10.0.0

PS. It's not a mistake that it is `v10.0.0`. We decided to skip all versions from `v1.0.0` to `v9.0.0` to avoid conflicts and confusion with CKEditor 4's versions.

**Comment by Reinmar** [comment]
I forgot to mention – together with CKEditor 5 v10.0.0 we released plugins enabling real-time collaborative editing in CKEditor 5 :)

You can find out more in https://docs.ckeditor.com/ckeditor5/latest/features/collaboration/overview.html

**Comment by Reinmar** [comment]
I updated the list in the first post with things that we introduced in:

* [CKEditor 5 v10.1.0](https://ckeditor.com/blog/CKEditor-5-v10.1.0-released/)
* [CKEditor 5 v11.0.0](https://ckeditor.com/blog/CKEditor-5-v11.0.0-released/)
* [CKEditor 5 v11.1.0](https://ckeditor.com/blog/CKEditor-5-v11.1.0-released/)

So this list now includes the following features:

* Tables support – https://github.com/ckeditor/ckeditor5-table
* Media embed support – https://github.com/ckeditor/ckeditor5-media-embed
* Autosave feature – https://ckeditor.com/docs/ckeditor5/latest/builds/guides/integration/saving-data.html#autosave-feature
* React integration – https://github.com/ckeditor/ckeditor5-react
* Angular integration – https://github.com/ckeditor/ckeditor5-angular
* Support for <kbd>Shift</kbd>+<kbd>Enter</kbd>
